plugins/cpu/cpu.go

```diff
@@ -13,31 +21,170 @@ const (
+		sum, count float64
+	)
+
+	for i := 0; i <= len(s.values)/2; i++ {
+		for _, v := range s.values[i : i+subSet] {
+			sum += v
+			count += 1
+		}
+		avgs = append(avgs, (sum / count))
+		sum = 0
+		count = 0
+	}
+
+	for _, v := range avgs {
+		sum += v
+		count += 1
+	}
+
+	return sum / count
+}
+
 type CPU struct {
+	mux           sync.Mutex
 	previous      []int
 	current       []int
 	currentTotal  int
 	previousTotal int
 	lastUpdate    time.Time
+	averages      []*sample
+	updaters      []metrics.Updater
+	slackURL      string
+}
 
-	updaters []metrics.Updater
+type Config struct {
+	Threshold         float64
+	SampleRate        time.Duration
+	ReportingInterval time.Duration
+	SlackURL          string
 }
 
-func New() *CPU {
-	return &CPU{
+func New(cfg *Config) *CPU {
+	// Determine the alpha factor for exponential moving average
+	alpha := cfg.SampleRate.Seconds() / cfg.ReportingInterval.Seconds()
+	var alerted bool
+
+	c := &CPU{
+		averages: []*sample{
+			userGauge:   newSample(alpha),
+			systemGauge: newSample(alpha),
+			idleGauge:   newSample(alpha),
+		},
 		updaters: []metrics.Updater{
 			userGauge:   metrics.NewGauge("cpu.user"),
 			systemGauge: metrics.NewGauge("cpu.system"),
 			idleGauge:   metrics.NewGauge("cpu.idle"),
 		},
+		slackURL: cfg.SlackURL,
+	}
+
+	go func() {
+		c.collect()
+		for range time.Tick(cfg.SampleRate) {
+			c.collect()
+			c.mux.Lock()
+			for _, i := range []int{userGauge, systemGauge, idleGauge} {
+				c.averages[i].add(c.rate(i))
+			}
+			avg := c.averages[userGauge].movingAvg()
+			if !alerted && avg > cfg.Threshold {
+				alerted = true
+				c.postSlack(fmt.Sprintf("[ALERT]: cpu.user average utilization %f is higher than %f", avg, cfg.Threshold))
+			} else if alerted && avg < cfg.Threshold {
+				alerted = false
+				c.postSlack("[RESOLVED]: cpu.user average utilization within threshold")
+			}
+			c.mux.Unlock()
+		}
+	}()
+
+	return c
+}
+
+func (c *CPU) postSlack(msg string) {
+	if len(c.slackURL) == 0 {
+		fmt.Printf("%s\n", msg)
+		return
 	}
+
+	client := &http.Client{Timeout: time.Second * 10}
+
+	//ip := cfg.hostIP
+	var ip string
+	if len(ip) == 0 {
+		var err error
+		ip, err = os.Hostname()
+		if err != nil {
+			ip = "unknown"
+		}
+	}
+	type message struct {
+		Text string `json:"text"`
+	}
+	m := message{Text: fmt.Sprintf("report from host %s\n%s", ip, msg)}
+
+	body, err := json.Marshal(&m)
+	if err != nil {
+		fmt.Printf("cpu: post stack notification: %v: %s\n", err, msg)
+		return
+	}
+
+	v := url.Values{}
+	v.Set("payload", string(body))
+	resp, err := client.PostForm(c.slackURL, v)
+	if err != nil {
+		fmt.Printf("cpu: post stack notification: %v: %s\n", err, msg)
+		return
+	}
+
+	defer resp.Body.Close()
+	io.Copy(ioutil.Discard, resp.Body)
 }
 
 func (c *CPU) Collect(b *metrics.Batch) {
-	c.collect()
+	c.mux.Lock()
 	for _, i := range []int{userGauge, systemGauge, idleGauge} {
-		c.updaters[i].Update(c.rate(i))
+		c.updaters[i].Update(c.averages[i].drain())
```

Comment:
The interplay between drain() and movingAvg() is a bit obscure. That weirdness goes away though if we just gather the expMovingAvg here instead of actually clearing anything.